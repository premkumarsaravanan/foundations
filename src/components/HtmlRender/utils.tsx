import * as React from 'react'
import { H4, H5 } from '../Typography'

export interface Element {
  type: string
  tagName?: string
  children?: Element[]
  attributes?: any
  content?: string
}

const sortContentType = (domItem: Element, index: number) =>
  domItem.type === 'text' ? domItem.content || null : rendererModule.sortTags(domItem, index)

const getChildren = (domTag: Element) =>
  domTag.children
    ? domTag.children.map((child: Element, index: number) => rendererModule.sortContentType(child, index))
    : null

const getAttributes = (domTag: Element, index: number) => {
  const attributes = domTag.attributes || {}
  return { ...attributes, style: {}, key: index }
}

const sortTags = (domTag: Element, index: number) => {
  const children = rendererModule.getChildren(domTag)
  const attributes = rendererModule.getAttributes(domTag, index)
  if (!children || !children.length) {
    return null
  }
  switch (domTag.tagName) {
    case 'p':
      return <p {...attributes}>{children}</p>
    case 'a':
      return <a {...attributes}>{children}</a>
    case 'b':
      return <b {...attributes}>{children}</b>
    case 'u':
      return <u {...attributes}>{children}</u>
    case 'ul':
      return <ul {...attributes}>{children}</ul>
    case 'ol':
      return <ol {...attributes}>{children}</ol>
    case 'li':
      return <li {...attributes}>{children}</li>
    case 'h1':
      return <H4 {...attributes}>{children}</H4>
    case 'h2':
      return <H5 {...attributes}>{children}</H5>
    case 'i':
      return <i {...attributes}>{children}</i>
    case 'strike':
      return <del {...attributes}>{children}</del>
    case 'blockquote':
      return <blockquote {...attributes}>{children}</blockquote>
    case 'pre':
      return <pre {...attributes}>{children}</pre>
    case 'hr':
      return <hr {...attributes} />
    default:
      return <div {...attributes}>{children}</div>
  }
}

const renderer = (domContent: Element[]) =>
  domContent.map((domItem: Element, index: number) => rendererModule.sortContentType(domItem, index))

export const rendererModule = {
  sortContentType,
  getChildren,
  getAttributes,
  sortTags,
  renderer
}
export default renderer
