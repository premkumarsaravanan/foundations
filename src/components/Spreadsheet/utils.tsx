import * as React from 'react'
import Papa from 'papaparse'
import { Cell } from './types'

/** Get max row and col of data */
export const getMaxRowAndCol = (data: Cell[][]): { maxRow: number; maxCol: number } => {
  const maxRow = data.length
  /* default to 0 */
  let maxCol = 0
  /* check every row to find max length of column */
  data.forEach(row => {
    const numberOfCurrentRowColumn = row.length
    if (maxCol < numberOfCurrentRowColumn) {
      maxCol = numberOfCurrentRowColumn
    }
  })
  return { maxRow, maxCol }
}

export const setCurrentCellValue = (
  cellData: string,
  data: Cell[][],
  row: number,
  col: number,
  setData: React.Dispatch<Cell[][]>
): void => {
  const newData = JSON.parse(JSON.stringify(data))
  newData[row][col].value = cellData
  setData(newData)
}

export const parseCsvFile = (file: File): Promise<Papa.ParseResult> =>
  new Promise(resolve => {
    Papa.parse(file, {
      complete: (results: Papa.ParseResult) => {
        resolve(results)
      }
    })
  })

/** convert to string to download */
export const unparseDataToCsvString = (parseResult: string[][]): string => Papa.unparse(parseResult)
/**
 * Convert from ParseResult to Cell[][]
 * Convert from [['value1', 'value2'], ['value3', 'value4']]
 * to [[{value: 'value1'}, {value: ''}], [{value: 'value3'}, {value: 'value4'}]]
 * for compatible reason
 */
export const convertToCompatibleData = (parsedResult): Cell[][] =>
  parsedResult.data.map(rowArray => rowArray.map(value => ({ value })))

/**
 * Convert back from Cell[][] to string[][]
 */
export const convertDataToCsv = (data: Cell[][]): string[][] => data.map(rowArray => rowArray.map(({ value }) => value))
