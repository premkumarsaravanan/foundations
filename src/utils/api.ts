import { getAccessToken } from './session'
import { CONTACTS_HEADERS } from '../constants/api'

export const initAuthorizedRequestHeaders = async () => ({
  ...CONTACTS_HEADERS,
  Authorization: `Bearer ${await getAccessToken()}`
})
