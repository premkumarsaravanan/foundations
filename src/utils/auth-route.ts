import { LoginType } from '@reapit/cognito-auth'
import Routes from '../constants/routes'

export function getAuthRouteByLoginType(loginType: LoginType) {
  switch (loginType) {
    case 'ADMIN':
      return Routes.ADMIN_LOGIN
    case 'DEVELOPER':
      return Routes.DEVELOPER_LOGIN
    default:
      return Routes.CLIENT_LOGIN
  }
}

export function getLoginTypeByPath(path: string) {
  switch (path) {
    case Routes.ADMIN_LOGIN:
      return 'ADMIN'
      break
    case Routes.DEVELOPER_LOGIN:
      return 'DEVELOPER'
      break
    default:
      return 'CLIENT'
  }
}
