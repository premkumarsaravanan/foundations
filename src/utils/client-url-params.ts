import { history } from '../core/router'
import { ClientParams } from '@/reducers/client'

export const addQuery = (query: ClientParams): string => {
  const currentParams = new URLSearchParams(history.location.search)
  for (let key in query) {
    if (query.hasOwnProperty(key)) {
      if (currentParams.has(key)) {
        currentParams.set(key, query[key])
      } else {
        currentParams.append(key, query[key])
      }
    }
  }
  const path = `${history.location.pathname}?${currentParams.toString()}`
  return path
}

export const removeQuery = (queries: Array<string>): string => {
  const currentParams = new URLSearchParams(history.location.search)
  queries.forEach(query => {
    if (currentParams.has(query)) {
      currentParams.delete(query)
    }
  })

  const path = `${history.location.pathname}?${currentParams.toString()}`
  return path
}

export const getParamsFromPath = (search: string): ClientParams => {
  const output = { page: 1, search: '', category: '' }
  const params = new URLSearchParams(search)
  if (params.has('page')) {
    const pageParam = Number(params.get('page'))
    output.page = !isNaN(pageParam) && pageParam > 0 ? pageParam : 1
  }
  if (params.has('search')) {
    output.search = params.get('search') || ''
  }
  if (params.has('category')) {
    output.category = params.get('category') || ''
  }

  return output
}

export const getParamValueFromPath = (search: string, paramName: string): string => {
  const params = new URLSearchParams(search)
  if (params.has(paramName)) {
    return params.get(paramName) || ''
  }

  return ''
}

export const hasFilterParams = (search: string): boolean => {
  const params = new URLSearchParams(search)
  return params.has('search') || params.has('category')
}
