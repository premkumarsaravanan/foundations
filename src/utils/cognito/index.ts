export * from './cognito'
export * from './types'
export * from './constants'
