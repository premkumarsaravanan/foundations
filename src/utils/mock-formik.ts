export const mockFormikAction = {
  setStatus: jest.fn(),
  setError: jest.fn(),
  setErrors: jest.fn(),
  setSubmitting: jest.fn(),
  setTouched: jest.fn(),
  setValues: jest.fn(),
  setFieldValue: jest.fn(),
  setFieldError: jest.fn(),
  setFieldTouched: jest.fn(),
  validateForm: jest.fn(),
  validateField: jest.fn(),
  resetForm: jest.fn(),
  submitForm: jest.fn(),
  setFormikState: jest.fn()
}
