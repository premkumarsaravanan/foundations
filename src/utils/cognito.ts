import store from '../core/store'
import { authLoginSuccess, authLogout } from '../actions/auth'
import { LoginSession, LoginType } from '../reducers/auth'
import fetcher from './fetcher'
import { COGNITO_API_BASE_URL, COGNITO_HEADERS } from '../constants/api'

export interface LoginParams {
  userName: string
  password: string
  loginType: LoginType
}

export interface RefreshParams {
  userName: string
  refreshToken: string
}

/**
 * Used to verify if an access token has expired by checking time now against token expiry
 */
export const tokenExpired = (expiry: number) => {
  // Adding a minute to avoid race conditions on the API and CPU clock innaccuracies
  const timeNowPlusMin = Math.round(new Date().getTime() / 1000) + 60
  return expiry < timeNowPlusMin
}

/**
 * Convenience method to log the user out when Cognito fails and ensure LoginSession is restored to
 * null in Redux
 */
export const logOutUser = (loginType: LoginType) => {
  store.dispatch(authLogout(loginType))
  return null
}

/**
 * Handles business logic to return an access token
 * If I don't have a session in Redux, logout
 * If I do have one in Redux and not expired, return that session
 * If not, refresh session
 * If that fails, logout
 */
export const getAccessToken = async (): Promise<string | null> => {
  const { loginSession, loginType } = store.state.auth

  if (!loginSession) {
    return logOutUser(loginType)
  }

  const sessionExpired = tokenExpired(loginSession.accessTokenExpiry)

  if (!sessionExpired) {
    return loginSession.accessToken
  }

  try {
    const refreshedSession = await refreshSession(loginSession)
    if (refreshedSession) {
      store.dispatch(authLoginSuccess(refreshedSession))
      return refreshedSession.accessToken
    }
  } catch (err) {
    console.error(err)
  }

  return logOutUser(loginType)
}

const refreshSession = async ({ userName, refreshToken }) => {
  const refreshedSession = (await fetcher({
    url: '/refresh',
    method: 'POST',
    api: COGNITO_API_BASE_URL,
    body: { userName, refreshToken },
    isPrivate: false,
    headers: COGNITO_HEADERS
  })) as LoginSession | undefined

  return refreshedSession
}

export default refreshSession
