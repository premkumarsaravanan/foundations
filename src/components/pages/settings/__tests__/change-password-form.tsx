import React from 'react'
import { shallow } from 'enzyme'
import {
  ChangePasswordForm,
  ChangePasswordFormProps,
  ChangePasswordValues,
  validateChangePasswordForm,
  mapPropsChangePassword,
  handleSubmitChangePassword
} from '../change-password-form'
import { mockFormikAction } from '@/utils/mock-formik'

jest.mock('@reapit/elements', () => ({
  fetcher: jest.fn().mockResolvedValue({ message: 'SUCCESS' })
}))

describe('ChangePasswordForm', () => {
  it('should match snapshot', () => {
    const mockProps = {
      isSubmitting: false,
      isValidating: false,
      isValid: true
    } as ChangePasswordFormProps
    const wrapper = shallow(<ChangePasswordForm {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should match snapshot', () => {
    const mockProps = {
      isSubmitting: true,
      isValidating: true,
      isValid: false
    } as ChangePasswordFormProps
    const wrapper = shallow(<ChangePasswordForm {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  describe('validateChangePasswordForm', () => {
    it('should not return errors', () => {
      const mockValues: ChangePasswordValues = {
        currentPassword: '123',
        password: '456',
        confirmPassword: '456'
      }
      const result = validateChangePasswordForm(mockValues)
      expect(result).toEqual({})
    })

    it('should return errors', () => {
      const mockValues: ChangePasswordValues = {
        currentPassword: '123',
        password: '456',
        confirmPassword: '4567'
      }
      const result = validateChangePasswordForm(mockValues)
      expect(result).toEqual({
        confirmPassword: 'Passwords do not match'
      })
    })
  })

  describe('mapPropsChangePassword', () => {
    it('should run correctly', () => {
      const result = mapPropsChangePassword()
      expect(result).toEqual({
        currentPassword: '',
        password: '',
        confirmPassword: ''
      })
    })
  })
  describe('handleSubmitChangePassword', () => {
    it('should call setSubmitting', () => {
      const mockValues: ChangePasswordValues = {
        currentPassword: '123',
        password: '456',
        confirmPassword: '456'
      }
      const mockForm = {
        ...mockFormikAction
      }
      const mockProps = {
        changePassword: jest.fn(),
        email: 'test@gmail.com'
      }
      handleSubmitChangePassword(mockValues, { ...mockForm, props: mockProps })
      expect(mockForm.setSubmitting).toBeCalledWith(true)
      expect(mockProps.changePassword).toBeCalled()
    })
  })
})
