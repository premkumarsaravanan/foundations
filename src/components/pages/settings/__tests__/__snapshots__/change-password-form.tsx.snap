// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordForm should match snapshot 1`] = `
<FormSection>
  <Form>
    <FormHeading>
      Change password
    </FormHeading>
    <FormSubHeading>
      Please complete the following fields to change your password. You will be automatically logged out for the changes to be applied
    </FormSubHeading>
    <Grid>
      <GridItem>
        <Input
          dataTest="current-password"
          id="currentPassword"
          labelText="Current Password"
          name="currentPassword"
          type="password"
        />
      </GridItem>
    </Grid>
    <Grid>
      <GridItem>
        <Input
          dataTest="password"
          id="password"
          labelText="Password"
          name="password"
          type="password"
        />
      </GridItem>
      <GridItem>
        <Input
          dataTest="confirmPassword"
          id="confirmPassword"
          labelText="Confirm Password"
          name="confirmPassword"
          type="password"
        />
      </GridItem>
    </Grid>
    <FlexContainerResponsive>
      <Button
        disabled={false}
        loading={false}
        type="submit"
        variant="primary"
      >
        Change Password
      </Button>
    </FlexContainerResponsive>
  </Form>
</FormSection>
`;

exports[`ChangePasswordForm should match snapshot 2`] = `
<FormSection>
  <Form>
    <FormHeading>
      Change password
    </FormHeading>
    <FormSubHeading>
      Please complete the following fields to change your password. You will be automatically logged out for the changes to be applied
    </FormSubHeading>
    <Grid>
      <GridItem>
        <Input
          dataTest="current-password"
          id="currentPassword"
          labelText="Current Password"
          name="currentPassword"
          type="password"
        />
      </GridItem>
    </Grid>
    <Grid>
      <GridItem>
        <Input
          dataTest="password"
          id="password"
          labelText="Password"
          name="password"
          type="password"
        />
      </GridItem>
      <GridItem>
        <Input
          dataTest="confirmPassword"
          id="confirmPassword"
          labelText="Confirm Password"
          name="confirmPassword"
          type="password"
        />
      </GridItem>
    </Grid>
    <FlexContainerResponsive>
      <Button
        disabled={true}
        loading={true}
        type="submit"
        variant="primary"
      >
        Change Password
      </Button>
    </FlexContainerResponsive>
  </Form>
</FormSection>
`;
